plugins/inputs/solrmetrics/solrmetrics.go

```diff
@@ -0,0 +1,413 @@
+				// Add requeremnted tags
+				copyStrMaps(addTags, series[newSerie].tags)
+				series[newSerie].tags["collection"] = collection
+				series[newSerie].tags["shard"] = shard
+				series[newSerie].tags["replica"] = replica
+			}
+			series[newSerie].fields[metricName] = v
+		}
+	}
+	res = []Serie{}
+	for _, v := range series {
+		// Add a serie, if fileds contain metrics
+		if len(v.fields) > 0 {
+			res = append(res, v)
+		}
+	}
+	return res
+}
+
+// SampleConfig returns sample configuration for this plugin.
+func (s *Solr) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the plugin description.
+func (s *Solr) Description() string {
+	return "Read stats from Solr servers and cores using Metrics API"
+}
+
+// NewSolr return a new instance of Solr
+func NewSolr() *Solr {
+	return &Solr{}
+}
+
+// Provide metrics urls
+func (s *Solr) metricsURLs(acc telegraf.Accumulator) ([]url.URL, error) {
+	if len(s.Prefixes) == 0 && len(s.Keys) == 0 {
+		return nil, fmt.Errorf("'keys' and 'prefix' are empty")
+	}
+	URLs := make([]url.URL, 0, len(s.Servers)*2)
+	for _, server := range s.Servers {
+		URLPrs := make(map[string]url.Values, 0)
+		if len(s.Keys) > 0 {
+			URLPrs["keys"] = url.Values{}
+		}
+		for _, key := range s.Keys {
+			URLPrs["keys"].Add("key", key)
+		}
+		if len(s.Prefixes) > 0 {
+			URLPrs["prefixes"] = url.Values{}
+			URLPrs["prefixes"].Set("prefix", strings.Join(s.Prefixes, ","))
+		}
+		// URL for a server
+		strURL := fmt.Sprintf("%s/solr%s", server, metricPath)
+		serverURL, err := url.Parse(strURL)
+		if err != nil {
+			acc.AddError(err)
+			continue
+		}
+		// Add common parameters and set parameters for the server URL
+		for _, v := range URLPrs {
+			v.Set("compact", compact)
+			v.Set("wt", wt)
+			serverURL.RawQuery = v.Encode()
+			URLs = append(URLs, *serverURL)
+		}
+
+	}
+	return URLs, nil
+}
+
+func (s *Solr) createHTTPClient() *http.Client {
+	// Set default value if the http timeout wasn't set in config
+	if s.Httptimeout == 0 {
+		s.Httptimeout = dftHTTPTimeout
+	}
+	tr := &http.Transport{
+		ResponseHeaderTimeout: s.Httptimeout * time.Second,
+	}
+	client := &http.Client{
+		Transport: tr,
+		Timeout:   s.Httptimeout * time.Second,
+	}
+	return client
+}
+
+func (s *Solr) gatherData(req *http.Request, m interface{}) error {
+	r, err := s.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer r.Body.Close()
+	if r.StatusCode != http.StatusOK {
+		return fmt.Errorf("solr: API responded with status-code %d, expected %d, url %s",
+			r.StatusCode, http.StatusOK, req.URL)
+	}
+	if err = json.NewDecoder(r.Body).Decode(m); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateReqs creates request for the each Solr node
+func (s *Solr) CreateReqs(acc telegraf.Accumulator) ([]*http.Request, error) {
+
+	urls, err := s.metricsURLs(acc)
+	if urls == nil {
+		return nil, fmt.Errorf("empty list of URLs to request the metrics. %s", err)
+	}
+
+	var reqs []*http.Request
```

Comment:
Can we please store this in the structure!? AFAICS the requests are not depending on data in `Gather()` and AFAIK `http.Request`s can be reused.